standards/continuous-delivery.md

```diff
@@ -0,0 +1,101 @@
+- **Relentlessly pursue continuous improvement.** The most important characteristic of high-performing teams is that they are never satisfied: they always strive to get better. High performers make improvement part of everybody's daily work.
+- **Everyone is responsible.** As we learned from Ron Westrum, in bureaucratic organizations teams tend to focus on departmental goals rather than organizational goals. Thus, development focuses on throughput, testing on quality, and operations on stability. However, in reality these are all system-level outcomes, and they can only be achieved by close collaboration between everyone involved in the software delivery process. A key objective for management is making the state of these system-level outcomes transparent, working with the rest of the organization to set measurable, achievable, time-bound goals for these outcomes, and then helping their teams work toward them.
+
+## Foundations for Continuous Delivery
+
+In order to implement continuous delivery, we must create the following foundations:
+
+- **Comprehensive configuration management.** It should be possible to provision our environments and build, test, and deploy our software in a fully automated fashion purely from information stored in version control. Any change to environments or the software that runs on them should be applied using an automated process from version control. This still leaves room for manual approvals—but once approved, all changes should be applied automatically.
+- **Continuous integration (CI).** Many software development teams are used to developing features on branches for days or even weeks. Integrating all these branches requires significant time and rework. Following our principle of working in small batches and building quality in, high-performing teams keep branches short-lived (less than one day's work) and integrate them into trunk/master frequently. Each change triggers a build process that includes running unit tests. If any part of this process fails, developers fix it immediately.
+- **Continuous testing.** Testing is not something that we should only start once a feature or a release is "dev complete." Because testing is so essential, we should be doing it all the time as an integral part of the development process. Automated unit and acceptance tests should be run against every commit to version control to give developers fast feedback on their changes. Developers should be able to run automated tests on their workstations in order to triage and fix defects. Testers should be performing exploratory testing continuously against the latest builds to come out of CI. No one should be saying they are "done" with any work until all relevant automated tests have been written and are passing.
```

Comment:
One other thing that comes to mind - in a world where the dev hits merge and the changes go out to production, even if we have comprehensive test coverage, it should still be on us to check our work. Ie, ideally, the person doesn't just hit 'merge' and move on, it should be up to us to make sure we didn't break something and that our change did what we expected. I don't think this should be too cumbersome either - could be just a quick refresh of a page or something most of the time